### ISSUE TITLE
Functional tests

### ISSUE THREAD

**Issue by ewjoachim** [pypi/warehouse#8616]
We've had 2 regressions lately that lead to reverts (#8591 and #8555) and [more](https://github.com/pypa/warehouse/pulls?q=is%3Apr+sort%3Aupdated-desc+in%3Atitle+revert+is%3Aclosed+-author%3Aapp%2Fdependabot-preview+) in the last months. While not all of those could have been caught before impacting production, I believe some may.

Our current testing strategy has its own qualities & defaults. On the one hand, we have 100% coverage, which helps limiting cases of incorrectly transcribing our thought into code. On the other hand, its large use of stubs, and its positioning that is relatively low-level doesn't give us that much confidence into how good the code we wrote fits real life. For this we would need higher-level tests.

I've seen a few tests in the test codebase labeled functional, and it doesn't strike me as something fundamentally different from the other kinds of tests we have.

I'm thinking about starting a higher-level test suite that would be more likely to catch "real-life" issues, by avoiding mocks & stubs where possible, and running on the complete stack. I'm ok to try and work on this but I'm also perfectly ok if someone else volunteers.

I've got a few thoughts of where I'd tend to go in the long run, but I don't want to get too precise before making a first working iteration. To be clear though, I'm talking about running a server, and then trying to run a scenario while interacting only through the HTTPS interface, as much as possible. This does not necessarily imply a headless browser, but at some point, we could decide to go this way. What I'd need though even for the first iteration would be an idea of a realistic and useful scenario.

The place where we had troubles recently were the authentication, so I'd tend to say:

- create a user
- verify email
- login (failure, success)
- add token
- upload package with credentials (failure, success)
- upload package with token (failure, success)
- delete package
- delete user

This should help us with #8598 in the short term and also to envision #7266

**Comment by ewjoachim** [comment]
Hm, somehow missed `tests/functional/legacy_api`. These resemble much closely something I had in mind :)

**Comment by di** [comment]
I'm all for more functional tests here. Might be nice to be able to split them from unit tests in the test suite (to run both in parallel).

I think https://github.com/pypa/warehouse/blob/master/tests/functional/legacy_api/test_basic.py is probably a good place to start with these additions.

**Comment by ewjoachim** [comment]
I'd like to do a test run of behaviour tests and see where this goes.

Here's a first draft of scenarii: https://gist.github.com/ewjoachim/ffe4d22fa6a5c626254ddb8b70c8eeec
Let's see if I manage to implement this in a sane & readable way :)

**Comment by miketheman** [comment]
I came across this spit behavior today, where I found that tests in the `unit/` folder make calls that access the database, and fail if the database is not running.
I know we've put in a fair amount of work to make the test suite work well via docker and make commands, so identifying when a test case is using the db in the scope of a file in the `unit/` namespace isn't immediately apparent today.

Is there continued value in having the split between unit and functional namespaces in the test suite, and if so, is there value in identifying and moving tests that use database access to functional? Our current development guidelines request unit tests https://warehouse.pypa.io/development/submitting-patches.html#tests and make no mention of functional.
Considering that the `functional` space mostly includes `webtest` tests, except for the `tests/functional/manage/test_views.py`, seems like there could be some good work on reorganizing the tests into the categories we expect them to be.

I can see a few patterns here:

- we don't worry too much about "pure" unit tests, and categorize anything that is not a web (browser or api) test as "unit"
- we tease apart unit vs functional (pure unit vs using a database), and add another category of "view" (better name tbd), and place items that interact with the warehouse UI or API service there
- something else entirely I haven't thought of

P.S. We also have our javascript Stimulus controller tests, which I think are out of scope for this conversation, since they have their own setup/test techniques. And if there's a desire to go full-UI-testing, I've played a bit with Cypress and got something working, and those could leverage BrowserStack if we wanted to go down that rabbit hole.

**Comment by woodruffw** [comment]
Triaging: I believe the functional testing story is now much better than it was at the time of last update (Jul 2022): https://github.com/pypi/warehouse/tree/main/tests/functional

**Comment by ewjoachim** [comment]
I'll follow your opinion on this. No problem closing if it's not an issue anymore

**Comment by miketheman** [comment]
Thanks gang! We have a few functional tests now, by no means as exhaustive as what @ewjoachim originally suggested, but enough coverage to feel relatively comfortable and adding them isn't too hard a task with the fixtures we have available.

See https://github.com/pypi/warehouse/tree/main/tests/functional for examples, and a specifc, longer one here: https://github.com/pypi/warehouse/blob/80c5fadf54a503ce6fe4751b23c1090329c6b6d2/tests/functional/forklift/test_legacy.py
